src/main/java/org/sonarlint/intellij/analysis/LocalFileExclusions.java

```diff
@@ -145,26 +143,124 @@ private ExcludeResult checkFileInSourceFolders(VirtualFile file, Module module)
     return ExcludeResult.notExcluded();
   }
 
-  public ExcludeResult canAnalyze(VirtualFile file, @Nullable Module module) {
-    if (powerSaveModeCheck.getAsBoolean()) {
-      return ExcludeResult.excluded("power save mode is enabled");
+  public Map<Module, Collection<VirtualFile>> retainNonExcludedFilesByModules(Collection<VirtualFile> files, boolean forcedAnalysis,
+    BiConsumer<VirtualFile, ExcludeResult> excludedFileHandler)
+    throws InvalidBindingException {
+    Map<Module, Collection<VirtualFile>> filesByModule = new LinkedHashMap<>();
+
+    for (VirtualFile file : files) {
+      checkExclusionsFileByFile(forcedAnalysis, excludedFileHandler, filesByModule, file);
+    }
+    if (!forcedAnalysis && !filesByModule.isEmpty()) {
+      // Apply server file exclusions. This is an expensive operation, so we call the core only once per module.
+      filterWithServerExclusions(excludedFileHandler, filesByModule);
     }
```

Comment:
If I'm not wrong, here we could check server exclusions even if the file was already excluded in `checkExclusionsFileByFile`. It sounds like a useless check

Reply:
`filesByModule` contains only non excluded files after `checkExclusionsFileByFile`.